File: townhall/db/agent.py
from typing import Dict, Optional, List, Union
from .agent_base import AgentBase

class Agent(AgentBase, table=True):
    class Config:
        property_set_methods = {"_name": "set_name"}

    _name: str
    system_message: str = None
    openai_id: str = None

    def __init__(
        self,
        name: str,
        system_message: str,
    ):
        super().__init__(name=name)
        self._name = name
        self.system_message = system_message

    def set_name(self, value):
        self.name = value

    def __setattr__(self, key, val):
        method = self.__config__.property_set_methods.get(key)
        if method is None:
            super().__setattr__(key, val)
        else:
            getattr(self, method)(val)

    @property
    def name(self):
        """Get the name of the agent."""
        return self._name

    def send(self, message: Union[Dict, str], recipient: "Agent", request_reply: Optional[bool] = None):
        """(Abstract method) Send a message to another agent."""

    async def a_send(self, message: Union[Dict, str], recipient: "Agent", request_reply: Optional[bool] = None):
        """(Abstract async method) Send a message to another agent."""

    def receive(self, message: Union[Dict, str], sender: "Agent", request_reply: Optional[bool] = None):
        """(Abstract method) Receive a message from another agent."""

    async def a_receive(self, message: Union[Dict, str], sender: "Agent", request_reply: Optional[bool] = None):
        """(Abstract async method) Receive a message from another agent."""

    def reset(self):
        """(Abstract method) Reset the agent."""

    def generate_reply(
        self,
        messages: Optional[List[Dict]] = None,
        sender: Optional["Agent"] = None,
        **kwargs,
    ) -> Union[str, Dict, None]:
        """(Abstract method) Generate a reply based on the received messages.

        Args:
            messages (list[dict]): a list of messages received.
            sender: sender of an Agent instance.
        Returns:
            str or dict or None: the generated reply. If None, no reply is generated.
        """

    async def a_generate_reply(
        self,
        messages: Optional[List[Dict]] = None,
        sender: Optional["Agent"] = None,
        **kwargs,
    ) -> Union[str, Dict, None]:
        """(Abstract async method) Generate a reply based on the received messages.

        Args:
            messages (list[dict]): a list of messages received.
            sender: sender of an Agent instance.
        Returns:
            str or dict or None: the generated reply. If None, no reply is generated.
        """

File: guru_api/app/models/agent.rb
class Agent < ApplicationRecord
  # serialize :skills, type: Array
  # serialize :file_ids, type: Array
  # serialize :metadata, type: Hash
  # {
  #   "id": "asst_abc123",
  #   "object": "assistant",
  #   "created_at": 1698984975,
  #   "name": "Math Tutor",
  #   "description": null,
  #   "model": "gpt-4",
  #   "instructions": "You are a personal math tutor. When asked a question, write and run Python code to answer the question.",
  #   "tools": [
  #     {
  #       "type": "code_interpreter"
  #     }
  #   ],
  #   "file_ids": [],
  #   "metadata": {}
  # }
  def convert_to_openai_object
    {
      "id": self.openai_id,
      "object": "assistant",
      "created_at": self.openai_created_at,
      "name": self.name,
      "description": self.description,
      "model": self.model,
      "instructions": self.system_message,
      "tools": self.tools,
      "file_ids": self.file_ids,
      "metadata": self.metadata
    }
  end

  # (Abstract method) Send a message to another agent.
  def send(message, recipient, request_reply = nil)
  end

  # (Abstract async method) Send a message to another agent.
  def a_send(message, recipient, request_reply = nil)
  end

  # (Abstract method) Receive a message from another agent.
  def receive(message, sender, request_reply = nil)
  end

  # (Abstract async method) Receive a message from another agent.
  def a_receive(message, sender, request_reply = nil)
  end

  # (Abstract method) Reset the agent.
  def reset
  end

  # (Abstract method) Generate a reply based on the received messages.
  def generate_reply(messages = nil, sender = nil, **kwargs)
  end

  # (Abstract async method) Generate a reply based on the received messages.
  def a_generate_reply(messages = nil, sender = nil, **kwargs)
  end
end
